haystack/components/preprocessors/markdown_header_splitter.py

```diff
@@ -0,0 +1,391 @@
+    """
+
+    def __init__(
+        self,
+        *,
+        infer_header_levels: bool = False,
+        page_break_character: str = "\f",
+        secondary_split: Literal["none", "word", "passage", "period", "line"] = "none",
+        split_length: int = 200,
+        split_overlap: int = 0,
+        split_threshold: int = 0,
+    ):
+        """
+        Initialize the MarkdownHeaderSplitter.
+
+        :param infer_header_levels: If True, attempts to infer and rewrite header levels based on content structure.
+            Useful for documents where all headers use the same level. Defaults to False.
+        :param page_break_character: Character used to identify page breaks. Defaults to form feed ("\f").
+        :param secondary_split: Optional secondary split condition after header splitting.
+            Options are "none", "word", "passage", "period", "line". Defaults to "none".
+        :param split_length: The maximum number of units in each split when using secondary splitting. Defaults to 200.
+        :param split_overlap: The number of overlapping units for each split when using secondary splitting.
+            Defaults to 0.
+        :param split_threshold: The minimum number of units per split when using secondary splitting. Defaults to 0.
+        """
+        self.infer_header_levels = infer_header_levels
+        self.page_break_character = page_break_character
+        self.secondary_split = secondary_split
+        self.split_length = split_length
+        self.split_overlap = split_overlap
+        self.split_threshold = split_threshold
+
+        # initialize secondary_splitter only if needed
+        if self.secondary_split != "none":
+            self.secondary_splitter = DocumentSplitter(
+                split_by=self.secondary_split,
+                split_length=self.split_length,
+                split_overlap=self.split_overlap,
+                split_threshold=self.split_threshold,
+            )
+
+    def _infer_header_levels(self, text: str) -> str:
+        """
+        Infer and rewrite header levels in the markdown text.
+
+        This function analyzes the document structure to infer proper header levels:
+        - First header is always level 1
+        - If there's content between headers, the next header stays at the same level
+        - If there's no content between headers, the next header goes one level deeper
+        - Header levels never exceed 6 (the maximum in markdown)
+
+        This is useful for documents where all headers are at the same level, such as
+        output from document conversion tools like docling.
+        """
+        logger.debug("Inferring and rewriting header levels")
+
+        # find headers
+        pattern = r"(?m)^(#{1,6}) (.+)$"
+        matches = list(re.finditer(pattern, text))
+
+        if not matches:
+            logger.info("No headers found in document; skipping header level inference.")
```

Comment:
I think we should either add identifying info about the document such as ID so we could more easily use this log for debugging purposes or I would drop it.